### DISCUSSION TITLE
Detect leaves on a tree (graph)

### DISCUSSION THREAD

**Discussion by joaquincabezas** [LukasZahradnik/PyNeuraLogic#47] [Q&A] [answered]
Hi! 

Suppose we have a tree (so G is connected and acyclic) as our input, and that the task is a graph classification (so finally we will be doing a graph readout).

We can consider different roles in this network, and therefore we may want to apply some computation based on this role. One of these roles is being a leaf in the tree, that is, a node which has only one incident edge (degree value of 1). Trying to define this leaf, I find issues:

```
template = Template()

template += R.not_leaf(V.X) <= (R.edge(V.X, V.Y), R.edge(V.X, V.Z),  R.special.alldiff(...))
template += R.leaf(V.X) <= (~R.not_leaf(V.X), R.node_feature(V.X))

template += R.predict <= R.leaf(V.X)
```

If I try with a simple line graph

<0>---<1>---<2>---<3>---<4>---<5>

I get the following errors:
```
(connectAllNeurons) [SEVERE] - Error: no input found for this neuron!!: terminal(2)
(connectAllNeurons) [SEVERE] - Error: no input found for this neuron!!: terminal(1)
(connectAllNeurons) [SEVERE] - Error: no input found for this neuron!!: terminal(3)
```

I understand there are no inputs precisely for the nodes where I don't want to put the "terminal" predicate.

If I don't negate the leaf:

```
template = Template()

template += R.not_leaf(V.X) <= (R.edge(V.X, V.Y), R.edge(V.X, V.Z), R.special.alldiff(...))
template += R.test_not_leaf(V.X) <= (R._not_leaf(V.X), R.node_feature(V.X))

template += R.predict <= R.test_not_leaf(V.X)
```

Then it works and select the nodes 1,2,3 and 4. In this case I put "not_leaf" as hidden explicitely (in the previous one, with the negation, I understand is always hidden).

![image](https://user-images.githubusercontent.com/1833086/222233617-bc0e0927-b0ae-441a-9804-a0eba2b93e2c.png)

Thanks for any advice!

**Comment by LukasZahradnik** [comment, marked as answer]
Hello,

Thanks for reporting this. I managed to reproduce it only when the dataset contains symmetrical edge predicates (i.e., it has both edge(X, Y) and edge(Y, X)).

I will try to look into it, but this seems like a bug within the grounder, so it's difficult to resolve it.

**Reply by joaquincabezas** [reply]
Thanks! I guess the symmetrical edge is the standard one for undirected edges, right?

I changed the network to one with only directed edges (from root -0- onwards) `0-->1-->2-->3-->4-->5`

In this case, then I need to change the rule of "not_leaf" to:

`template += R.not_leaf(V.X) <= (R.edge(V.X, V.Y), R.edge(V.Z, V.X),  R.special.alldiff(...))` to reflect the direction of the edges.

Then I have the same results, I receive the errors for the negated one, but it works for the second one.

Maybe the problem is with the rule? I understand it as:

for each node X: 
- V.X = X
- Find edge V.X->V.Y and V.Z->V.X (with different value for X, Y, Z)
- Only attach "not_leaf" if BOTH edges are found

So for node 0, as we are only connected to 1 (edge(0,1)), this rule is not triggered so no "not_leaf" attached to node 0
For nodes 1 we can find V.Y=0 and V.Z=2, and so on for nodes 2,3,4
For node 5, we are only connected through edge(4,5),  this rule is not triggered so no "not_leaf" attached to node 0

**Reply by joaquincabezas** [reply]
something that can be relevant:

The template that is working generates a correct drawing:

![image](https://user-images.githubusercontent.com/1833086/222278870-1ae664e8-4e48-4215-a9f4-a9f45a6dd803.png)

While the one that is not working (negated) has the following output:

![image](https://user-images.githubusercontent.com/1833086/222279002-d0bba8d6-be9d-4213-bc9b-aa7ffd3fc078.png)

**Reply by LukasZahradnik** [reply]
Ah, right. It doesn't work even on a simpler instance.

@GustikS Any idea how to fix the problem? I tried to run it directly in Java, and it does not work properly.

Template:
```
not_leaf(X) :- edge(X, Z).
leaf(X) :- !*not_leaf(X), node_feature(X).
predict(X) :- leaf(X).
```

Examples:
```
1.0 edge(1, 2),
1.0 node_feature(1),
1.0 node_feature(2).
```

Queries:
```
1.0 predict(X).
```

For some reason, `leaf(X)` has two groundings `leaf(1)` and `leaf(2)` - it should only have `leaf(2)`.
Also, should the drawing of the template be disconnected as @joaquincabezas  shows? It seems just like a visual issue because the `liftedHashCode` is negated for negated literals.

**Reply by GustikS** [reply]
Hi, sorry I'm super busy ATM, but I will try to look at it on monday! 
Yes the drawing is just a visual issue, but the negation, which is a recent thing, and generally tricky in logic programming, is probably going to be problematic in our case too, I'm afraid...

**Reply by LukasZahradnik** [reply]
@GustikS I tried some things and managed to fix the issue. It doesn't break anything (according to passing tests), although I'm not really sure about its correctness.
I changed the [HornClause::getNegatedLiterals](https://github.com/GustikS/NeuraLogic/blob/master/Logic/src/main/java/cz/cvut/fel/ida/logic/HornClause.java#L126) to the following:

```java
    public Set<Literal> getNegatedLiterals() {
        Set<Literal> literalSet = new HashSet<>(body.literals());

        // ### Added
        for (Literal l : body.literals()) {
            if (l.isNegated()) {
                literalSet.add(new Literal(tupleNotInPredicateName(l.predicate()), false, l.arguments()));
            }
        }
        // ###

        literalSet.add(head.negation());
        return literalSet;
    }
```

So, for each negated literal, I added another special NotInTuple literal. The same is done for the negated head literal.

**Reply by GustikS** [reply]
Hi, yeah, that's actually smart! But we should definitely clarify what is happening here in the first place. The thing is that this is indeed a principled issue with negation in logic programming (and not a bug of the framework). When people write a logic template, they probably imagine it is evaluated top->down (& left->right) as in Prolog. But this is Datalog and that is not the case (Prolog is too slow for this). All the rules are evaluate *at once*, and their order should not matter (if it was not for this extra trickery of ours with the tupleNotInPredicateName and other side effects), and this is repeated until we find the least Herbrand model. So actually it is correct that
```
leaf(X) :- !*not_leaf(X), node_feature(X).
```
derives both `leaf(1)` & `leaf(2)`, since none of them satisfies not_leaf(X) at the beginning of grounding (in the later iterations `leaf(2)` is derived, but that does not matter since both have already been proved). 

Hence the solution here can be to enforce the ordering, so that the second rule fires only after the first rule has already proved `not_leaf(1)`. This is what the tupleNotInPredicateName does - it checks against the *current* Herbrand model within the solver. And luckily, we kind of keep the order of the rules as given by the user, so this kinda works. 
But it would be good to keep in mind that this is sort of an artefact and not very general/stable - as soon as we combine that with recursion, things could get messy pretty fast :D I think that we should mark the recent addition of negation as experimental or something, so that people don't go too crazy with it... it should be safe to use negation with the relations from the data (example file), which are explicitly given at the start so there the precedence is clear. But using negation on the derived predicates then requires deeper thinking on the user side about the internals of the proving process (that are hidden), which we should probably warn about and make clear in documentation at some point...

**Reply by LukasZahradnik** [reply]
Aha, thanks. I was kind of assuming that there was an ordering (because of the dependency on a different rule).

> And luckily, we kind of keep the order of the rules as given by the user, so this kinda works.

If users do not provide the program in the "correct" order, then it will not work, which isn't that great.

> I think that we should mark the recent addition of negation as experimental or something, so that people don't go too crazy with it

Agree. I will add a note about negation to the documentation once this issue is at least partially resolved. Right now, it's not documented at all, so it shouldn't be considered as something supported.

**Reply by GustikS** [reply]
Yeah, what we should ideally do is to restrict the templates to so-called "stratified" logic programs, which is exactly about this issue with negation. However that's probably far beyond what we can expect from users, and I don't know how to test for stratification automatically (never tried, but likely impossible to do in general).
But maybe some approximate rule ordering/partitioning/stratification could be derived from the GraphTemplate to check whether the negation is likely going to cause troubles or not...

**Comment by joaquincabezas** [comment]
Thanks for the explanation, it is eye-opening in relation to the negation question and why is it an principled issue in Datalog (and yes I was kind of thinking of an "magic ordering"). 

A workaround I put in place while discussing this: simply moving some of the rules out of the template to a previous stage of "preprocessing". This of course have been done outside the language of Neuralogic (basically with networkx, then converting the dataset back to Neuralogic). 

Although this technically works, it makes the pipeline less elegant and somehow spoils the idea of working within a logic framework, so I was thinking... Maybe a preprocessing stage where rules modify the facts? I am thinking of something in the spirit of this MsC thesis: [Preprocessing in Inductive Logic Programming by Brad Hunter](https://arxiv.org/abs/2112.12551)

I see like splitting the template into two stages, one which is super simple and only creates new predicates on the dataset (the preprocessing, ordered top down) and the current one where we do exploit the LRNNs. In the preprocessing we will just save the grounding into the examples.

Is there any strategy other than this that you can think of? Or maybe I am going too complex for an issue that can be solved in another way!

**Reply by GustikS** [reply]
Hi @joaquincabezas, I agree!

Well, first to be clear, I think that the solution proposed by @LukasZahradnik is going to work in your case (detecting leaves), as well as many other cases of stratified logic programs (so let's just try that!). It will detect the negation in body and automatically "turn on" the ordering as probably imagined by the user. It's just that there is generally this odd thing with negation in logic programming to be careful about (and it might be difficult to realize/detect it).

I think that moving things that are not related to learning to some preprocessing stage should be a best practice in general - not just because of this issue (which can be resolved), but also performance (as forward/backprop will unnecessarily go through these operations repeatedly every epoch). Generally, if there are too many rules near the "input" that do not have any weights attached, you are probably unnecessarily overloading the model with preprocessing - and "moving that to the examples" (calculating this only once at the start) is a good idea (actually there is a related LiftedExample class that may also contain rules, so we might revive that).

I agree that it would be more elegant to do this altogether in logic, and actually the grounder is quite powerful in extracting relational patterns, including (stratified) negations, so it might even be better than using external library (except for the unstratified negation and other advanced cases, where it might be better to preprocess with some ad-hoc procedural semantics). But we should be careful about modyfing the user's input example facts (as the template can be changing). So just marking the "preprocessing" rules as such in the template should be fine I guess...
@LukasZahradnik I imagine something along the lines of the current *hidden* modifier, but on the level of whole rules maybe...?

**Reply by LukasZahradnik** [reply]
I don't think I fully understand the purpose of the "preprocessing" rules.

If a user has some preprocessing in their template, won't it be pruned and compressed? So forward/backward will not go through those operations?

**Reply by GustikS** [reply]
It might get pruned and/or compressed in some cases, depending on situation, but not in general, I think. But yeah, it would be just a form of pruning, where we would prune out the preprocessing subgraph (not just a chain as it is now), and make its end nodes the new fact neurons (hence effectively skipping it in training). 

But let's keep that as a future todo, and right now we can just close this with your previous proposal, I guess?

**Reply by joaquincabezas** [reply]
I've marked it as an answer and I am opening a [feature request](https://github.com/LukasZahradnik/PyNeuraLogic/issues/50) so we can track it. Thanks!!!

**Reply by LukasZahradnik** [reply]
@joaquincabezas Thanks for opening the issue.
@GustikS Should I open PR for adding the tupleNotIn extra predicate, or will you add it?

**Reply by GustikS** [reply]
@LukasZahradnik I will add it. I expect it to resolve most issues with negation like this one (hopefully).
@joaquincabezas just to be clear, the pruning feature is not directly related to this issue with negation, it's really just about performance. We will look at it in future (not in few months at least). It would definitely help if you could provide some of your templates in the meantime, so that we can include them for benchmarking/tuning the feature, thanks!

**Comment by GustikS** [comment]
I realized that the (global) stratification w.r.t. negation can actually be checked quite easily, so I updated the backend respectively. 
It now emits a severe warning if there is an unstratified negation detected in the template.
https://github.com/GustikS/NeuraLogic/commit/7c1f2940f994bb1f7e101207db946e80c3abef7e
@LukasZahradnik just please make sure that the frontend will not skip this new check.
I also updated the template drawing to remain connected when there are negated literals...
Hence I consider this negation topic as closed (hopefully) :)

**Comment by LukasZahradnik** [comment]
@joaquincabezas Hello, the fix for the negation is now published in the 0.7.5 version.

**Reply by joaquincabezas** [reply]
Good, I've just tested it and works :) Thanks

![image](https://user-images.githubusercontent.com/1833086/231244067-f53b8ef8-12f3-484f-8d82-59363d116299.png)

ps. also noticed the arrows now point in the direction of the data flow, good to see that also :)
